bai2/writers.py

```diff
@@ -0,0 +1,274 @@
+from collections import OrderedDict
+
+from .models import \
+    Bai2File, Bai2FileHeader, Bai2FileTrailer, \
+    Group, GroupHeader, GroupTrailer, \
+    AccountIdentifier, AccountTrailer, Account, \
+    TransactionDetail
+from .utils import write_date, write_time, convert_to_string
+from .constants import CONTINUATION_CODE
+from .conf import settings
+
+
+class BaseWriter(object):
+
+    def __init__(self, obj):
+        self.obj = obj
+
+    def write(self):
+        raise NotImplementedError()
+
+
+class BaseSectionWriter(BaseWriter):
+    model = None
+    header_writer_class = None
+    child_writer_class = None
+    trailer_writer_class = None
+
+    def write(self):
+        header = self.header_writer_class(self.obj.header).write()
+
+        children = []
+        for child in self.obj.children:
+            children.append(self.child_writer_class(child).write())
+
+        self.obj.update_totals()
+        self.obj.trailer.number_of_records = len(
+            '\n'.join([header] + children).split('\n')) + 1
```

Comment:
Yeah, it's not great.

It's because of continuation records. Each record writer writes itself out as a single complete string, including any continuation records, whereas the record count at the end needs to count continuation records separately.

Making each writer return a list of lines instead of just a string is something that I considered doing but didn't, because of reasons. I can't remember what they were. I might try again.